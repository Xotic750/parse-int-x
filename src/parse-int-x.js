import NAN from 'nan-x';
import toStr from 'to-string-x';
import trimLeft, {trimLeft2016} from 'trim-left-x';

const nativeParseInt = parseInt;
/**  @type {Function} */
const castNumber = (0).constructor;
// noinspection JSPotentiallyInvalidConstructorUsage
const {charAt} = '';
const hexRegex = /^[-+]?0[xX]/;
const {test} = hexRegex;

/**
 * This method parses a string argument and returns an integer of the specified
 * radix (the base in mathematical numeral systems). (ES2016).
 *
 * @param {string} [string] - The value to parse. If the string argument is not a
 *  string, then it is converted to a string (using the ToString abstract
 *  operation). Leading whitespace in the string argument is ignored.
 * @param {number} [radix] - An integer between 2 and 36 that represents the radix
 *  (the base in mathematical numeral systems) of the above mentioned string.
 *  Specify 10 for the decimal numeral system commonly used by humans. Always
 *  specify this parameter to eliminate reader confusion and to guarantee
 *  predictable behavior. Different implementations produce different results
 *  when a radix is not specified, usually defaulting the value to 10.
 * @throws {TypeError} If target is a Symbol or is not coercible.
 * @returns {number} An integer number parsed from the given string. If the first
 *  character cannot be converted to a number, NaN is returned.
 */
export function parseInt2016(string, radix) {
  const str = trimLeft2016(toStr(string));

  return nativeParseInt(str, castNumber(radix) || (test.call(hexRegex, str) ? 16 : 10));
}

/**
 * This method parses a string argument and returns an integer of the specified
 * radix (the base in mathematical numeral systems). (ES2018).
 *
 * @param {string} [string] - The value to parse. If the string argument is not a
 *  string, then it is converted to a string (using the ToString abstract
 *  operation). Leading whitespace in the string argument is ignored.
 * @param {number} [radix] - An integer between 2 and 36 that represents the radix
 *  (the base in mathematical numeral systems) of the above mentioned string.
 *  Specify 10 for the decimal numeral system commonly used by humans. Always
 *  specify this parameter to eliminate reader confusion and to guarantee
 *  predictable behavior. Different implementations produce different results
 *  when a radix is not specified, usually defaulting the value to 10.
 * @throws {TypeError} If target is a Symbol or is not coercible.
 * @returns {number} An integer number parsed from the given string. If the first
 *  character cannot be converted to a number, NaN is returned.
 */
const parseInt2018 = function parseInt2018(string, radix) {
  const str = trimLeft(toStr(string));

  if (charAt.call(str, 0) === '\u180E') {
    return NAN;
  }

  return nativeParseInt(str, castNumber(radix) || (test.call(hexRegex, str) ? 16 : 10));
};

export default parseInt2018;
