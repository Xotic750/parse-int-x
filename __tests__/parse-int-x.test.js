import $parseInt from 'src/parse-int-x';

const hasSymbol = typeof Symbol === 'function' && typeof Symbol('') === 'symbol';
const ifSymbolIt = hasSymbol ? it : xit;

describe('$parseInt', function() {
  const ws2016 =
    '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u180E\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028\u2029\uFEFF';
  const ws2018 =
    '\x09\x0A\x0B\x0C\x0D\x20\xA0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028\u2029\uFEFF';

  describe('basic', function() {
    it('$parseInt is a function', function() {
      expect.assertions(1);
      expect(typeof $parseInt).toBe('function');
    });
  });

  describe('$parseInt', function() {
    it('accepts a radix', function() {
      expect.assertions(35);
      for (let i = 2; i <= 36; i += 1) {
        expect($parseInt('10', i)).toBe(i);
      }
    });

    it('defaults the radix to 10 when the number does not start with 0x or 0X', function() {
      expect.assertions(4);
      ['01', '08', '10', '42'].forEach(function(str) {
        expect($parseInt(str)).toBe($parseInt(str, 10));
      });
    });

    it('defaults the radix to 16 when the number starts with 0x or 0X', function() {
      expect.assertions(2);
      expect($parseInt('0x16')).toBe($parseInt('0x16', 16));
      expect($parseInt('0X16')).toBe($parseInt('0X16', 16));
    });

    it('ignores leading whitespace', function() {
      expect.assertions(5);
      expect($parseInt('  0x16')).toBe($parseInt('0x16', 16));
      expect($parseInt('  42')).toBe($parseInt('42', 10));
      expect($parseInt('  08')).toBe($parseInt('08', 10));
      expect($parseInt(`${ws2018}08`)).toBe($parseInt('08', 10));
      expect($parseInt(`${ws2018}0x16`)).toBe($parseInt('0x16', 16));
    });

    it('leading whitespace 2018 should return NaN', function() {
      expect.assertions(2);

      expect(Number.isNaN($parseInt(`${ws2016}08`))).toBe(true);

      expect(Number.isNaN($parseInt(`${ws2016}0x16`))).toBe(true);
    });

    it('defaults the radix properly when not a true number', function() {
      expect.assertions(2);
      const fakeZero = {
        valueOf() {
          return 0;
        },
      };

      expect($parseInt('08', fakeZero)).toBe($parseInt('08', 10));
      expect($parseInt('0x16', fakeZero)).toBe($parseInt('0x16', 16));
    });

    it('allows sign-prefixed hex values', function() {
      expect.assertions(4);
      expect($parseInt('-0xF')).toBe(-15);
      expect($parseInt('-0xF', 16)).toBe(-15);
      expect($parseInt('+0xF')).toBe(15);
      expect($parseInt('+0xF', 16)).toBe(15);
    });

    it('naN parsing', function() {
      expect.assertions(4);
      expect($parseInt()).toBeNaN();
      expect($parseInt(undefined)).toBeNaN();
      expect($parseInt(null)).toBeNaN();
      expect($parseInt(NaN)).toBeNaN();
    });

    it('should throw if target is not coercible', function() {
      expect.assertions(1);
      expect(function() {
        $parseInt(Object.create(null));
      }).toThrowErrorMatchingSnapshot();
    });

    ifSymbolIt('should throw for Symbol', function() {
      expect.assertions(2);

      const sym = Symbol('foo');
      expect(function() {
        $parseInt(sym);
      }).toThrowErrorMatchingSnapshot();

      const symObj = Object(sym);
      expect(function() {
        $parseInt(Object(symObj));
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
