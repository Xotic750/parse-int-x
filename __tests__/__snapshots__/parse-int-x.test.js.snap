// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`$parseInt $parseInt should throw for Symbol 1`] = `"Cannot convert a Symbol value to a string"`;

exports[`$parseInt $parseInt should throw for Symbol 2`] = `"Cannot convert a Symbol value to a string"`;

exports[`$parseInt $parseInt should throw if target is not coercible 1`] = `"Cannot convert object to primitive value"`;
